### ISSUE TITLE
Statement time out not working as expected

### ISSUE THREAD

**Issue by kartikarora2** [jasync-sql/jasync-sql#338]
statementTimeout introduced in release r2dbc-spi [v0.9.0.M2](https://github.com/r2dbc/r2dbc-spi/releases/tag/v0.9.0.M2) is not working. We are using 0.9.0.RELEASE version of r2dbc-spi and using jasync-sql implementing library. We have set the statement timeout using connection factory options as below.

<img width="1646" alt="Screenshot 2022-12-07 at 12 28 28 AM" src="https://user-images.githubusercontent.com/119792820/205998260-d987d961-c237-42eb-ab61-ae9bd185bead.png">

To test statementTimeout we had set set it's value as 1ms. But our query ran for around 12ms, ideally it should have timeout because of the statement timeout parameter and thrown R2dbcTimeoutException. But the query ran successfully.

We also checked the JasyncClientConnection class where the statement timeout is being set for every session, but the statementtimeout values we had set using connection factory options, does not seem to work.

https://github.com/jasync-sql/jasync-sql/blob/master/r2dbc-mysql/src/main/java/JasyncClientConnection.kt#L76-L77

**Comment by oshai** [comment]
Hi,
Statement timeout dictates [MAX_EXECUTION_TIME](https://dev.mysql.com/doc/refman/5.7/en/optimizer-hints.html#optimizer-hints-execution-time) in mysql, which works for 
```
The MAX_EXECUTION_TIME hint is permitted only for SELECT statements. It places a limit N (a timeout value in milliseconds) on how long a statement is permitted to execute before the server terminates it:
```

So this is a server side timeout that works only for select statements. We can also configure client side timeout on the connection (It might throw a different exception) that jasync already support. From [spi](https://github.com/r2dbc/r2dbc-spi/issues/173) this looks like an reasonable behavior of the driver.

**Comment by sushilnayak** [comment]
Ok, i made a mistake, read through the code and realised **setStatementTimeout** never gets called in the create() method of JasyncConnectionFactory. So that means if i enter StatementTimeout parameter during connection configuration, it won't work. We could 

So, based on my prev comment on how Statement Timeout works, it looks like we should not even use **setStatementTimeout** because of the way the Mysql would respond backs when timeout occurs i.e. value of 1.

**innodb_lock_wait_timeout** also does not look like the right solution for sql query which are not locking row and i won't be writing a SELECT statement with FOR UPDATE to lock my select statement. 

So it means, the user has to create a time based wrapper function which wraps databaseClient call. Like i mentioned in my previous comment.

**Comment by oshai** [comment]
ok, based on what you found I suggest two fixes:
- make the timeout value in `set MAX_EXECUTION_TIME` in millis.
- on the factory call that method if such option exists.

In addition, I found out that we use another option to configure timeout `QUERY_TIMEOUT ` see [code](https://github.com/jasync-sql/jasync-sql/blob/254c30e91502a02b682f024f4387738abd8aa4a6/r2dbc-mysql/src/main/java/MysqlConnectionFactoryProvider.kt#L77).  

I suspect that this was added as an option for the driver before the spi supported that, so additional option is to change QUERY_TIMEOUT to use the spi option STATEMENT_TIMEOUT.

Let me know what you think.

**Comment by sushilnayak** [comment]
> ok, based on what you found I suggest two fixes:
> 
> - make the timeout value in set MAX_EXECUTION_TIME in millis.
> - on the factory call that method if such option exists.

point 1 is fine, we can fix that, but point 2 will not help in solving StatementTimeout problem. Because event if we do implement, result in case of timeout would be 1 ( also explained in my previous comment)

> `QUERY_TIMEOUT` to use the spi option `STATEMENT_TIMEOUT`.

got it.. i have tried it at my end, based on what i read so far, i think this should work. It's pretty much the wrapper solution kind of thing.

**Comment by oshai** [comment]
yes, PR will be great.

**Comment by kartikarora2** [comment]
@oshai, I had raised a PR for fixing the statement timeout issue. Can you please review it.

**Comment by oshai** [comment]
Fixed in 2.1.9
